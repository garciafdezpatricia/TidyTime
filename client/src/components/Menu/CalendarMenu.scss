@use "./../../../styles/variables" as *;


.calendar-menu-popup {
    font-size: .9rem;
    position:absolute;
    top: 3.1rem;
    left: 1rem;
    background-color: $white;
    height: calc(100% - 3.5rem);
    width: 23rem;
    z-index: 5;
    //box-shadow: -.7rem .2rem 3.5rem -.3rem rgb(147, 151, 147);
    border-radius: .2rem;
    border: .1rem solid $light-grey;
    padding: 1rem;
    overflow-y: auto;
    box-sizing: border-box;
    &:hover {
        border-color: $medium-grey;
    }

    @media screen and (min-width: 320px) and (max-width: 500px) {
        width: 90%;
    }

    .calendar-menu-events {
        display: flex;
        flex-direction: column;
        font-size: .8rem;

        .calendar-menu-buttons {
            display: flex;
            gap: .6rem;

            .confirm-btn {
                display: flex;
                align-items: center;
                border: none;
                background-color: $primary-green;
                color: $white;
                padding: .1rem .5rem;
                border-radius: .3rem;
                font-size: .8rem;
                width: fit-content;
                margin-bottom: .2rem;
            }
        }
    }

    .create-new-event, .import-google-btn {
        display: flex;
        align-items: center;
        gap: .4rem;
        border-radius: .3rem;
        border: none;
        width: fit-content;
        padding: .1rem .5rem;
        font-size: .8rem;
        color: $black;
    }

    .create-new-event {
        margin-bottom: .2rem;
        &:hover {
            background-color: $hover-green;
        }
    }

    .import-google-btn {
        &:hover {
            background-color: $light-blue;
        }
    }

    .calendar-menu-google {
        padding-top: 2rem;
        font-size: .8rem;

        .calendar-menu-google-header {
            display: flex;
            gap: .5rem;
        }

        .calendar-menu-google-body {
            display: flex;
            flex-direction: column;
            gap: .5rem;
        }

        .calendar-menu-google-calendars {
            list-style-type: none;
            padding-top: 1rem;

            .calendar-item:not(:last-child) {
                margin-bottom: .8rem;
            }

            .calendar-item {
                display: flex;
                align-items: center;
                gap: .5rem;

                p {
                    padding-inline: .5rem;
                    border-radius: .4rem;
                }
            }

            .calendar-item-sync-btn {
                height: 100%;
                width: fit-content;
                border: none;
                outline: none;
                background-color: transparent;
                padding-inline: .2rem;
                &:hover {
                    outline: .1rem solid $light-blue;
                }
            }
        }
    }
}