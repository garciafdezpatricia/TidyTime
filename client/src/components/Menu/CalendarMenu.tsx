import { useEffect, useRef, useState } from "react";
import { BsCalendarEventFill } from "react-icons/bs";
import { Icon } from "../Icon/Icon";
import NewEventForm from "../Event/NewEventForm";
import { useEventContext } from "../Context/EventContext";
import { uuid } from "uuidv4";
import { useClickAway } from "@uidotdev/usehooks";
import { IoMdSync } from "react-icons/io";
import { BsFillCalendarPlusFill } from "react-icons/bs";
import { CalendarItem } from "@/src/task/Scheme";
import { useGoogleContext } from "../Context/GoogleContext";
import toast from "react-hot-toast";
import { PiWarningOctagonFill } from "react-icons/pi";
import { IoMenu } from "react-icons/io5";

export default function CalendarMenu() {
    // event context utils
    const { setEvents } = useEventContext()
    const { calendars, setCalendars } = useGoogleContext();

    const [createNewEvent, setCreateNewEvent] = useState(false);
    const [selectedColor, setSelectedColor] = useState("");
    const [menuOpened, setMenuOpened] = useState(false);

    const titleRef = useRef(null);
    const infoRef = useRef(null);
    const fromDateRef = useRef(null);
    const toDateRef = useRef(null);
    const ref = useClickAway(() => {
        setMenuOpened(false);
    })

    /**
     * Auxiliary function to check if the required fields to create the event are filled 
     * @param title containing the title of the event
     * @param from containing the start date of the event
     * @param to containing the end date of the event
     * @returns true if all fields are completed. False otherwise.
     */
    function areFieldsCompleted(title:string, from:string, to:string) {
        let result = true;
        if (title === "") {
          // @ts-ignore
          titleRef.current.style.borderColor = "#e13535";
          result = false;
        }
        if (from === "") {
          // @ts-ignore
          fromDateRef.current.style.borderColor = "#e13535";
          result = false;
        }
        if (to === "") {
          // @ts-ignore
          toDateRef.current.style.borderColor = "#e13535";
          result = false;
        }
        return result;
    }

    /**
     * Handler for the creation of the event. If the event is correctly created, the creation of the event modal is closed
     */
    const handleCreateEvent = () => {
        if (createEvent()) {
            setMenuOpened(false);
        }
    }

    /**
     * Handler for color changes.
     */
    const handleColorChange = (color:string) => {
        setSelectedColor(color);
    }
    
    /**
     * Creates a new event from the value of the fields. If all required fields are filled, the creation action takes place.
     * @returns 
     */
    const createEvent = () => {    
        // @ts-ignore
        const title = titleRef.current.value; const info = infoRef.current.value; 
        // @ts-ignore
        const from = fromDateRef.current.value; const to = toDateRef.current.value; 
        if (areFieldsCompleted(title, from, to)) {
            // need to create the event dates with new Date() for them to be displayed on the calendar
            setEvents((prev) => 
            [...prev, {start: new Date(from), end: new Date(to), title: title, desc: info, color: selectedColor, eventId: uuid()}]
            );
            return true;
        }
        return false;
    }

    const getGoogleCalendars = () => {
        fetch('http://localhost:8080/google/calendar/list', {
        method: 'GET',
        credentials: 'include',
        })
        .then(response => response.json())
        .then(data => {
            if (data.status === 'success') {
                const calendars: CalendarItem[] = [];
                data.value.map((calendar:any) => {
                    calendars.push({id: calendar.id, name: calendar.summary, color: calendar.backgroundColor})
                });
                setCalendars(calendars);
            } else {
                console.error(data.value);
                toast.error("Couldn't import the calendars.\nPlease log in with your Google account and try again.", {
                    position: "top-center",
                    duration: 6000,
                    icon: <></>,
                    style: {
                        textAlign: "center"
                    }
                })
            }
        })  
        .catch(error => console.error('Error:', error));
    }

    return (
        // @ts-ignore
       <div ref={ref}>
            <button className="calendar-menu-icon" onClick={() => setMenuOpened(!menuOpened)} >
                <IoMenu size={"1.4rem"} style={{color: "#3E5B41", backgroundColor: "transparent", borderRadius: "0.4rem"}}/>
            </button>
            {
                menuOpened &&
                // @ts-ignore
                <article className="calendar-menu-popup">
                    <section className="calendar-menu-events">
                        <h3>Add event</h3>
                        <hr/>
                        <p>Create a new event and add it to your calendar.</p>
                        <section className="calendar-menu-buttons">
                            <button onClick={() => setCreateNewEvent(!createNewEvent)} className="create-new-event">
                                <BsCalendarEventFill size={".9rem"}/>
                                New event
                            </button>
                            { createNewEvent && <button className="confirm-btn" onClick={() => handleCreateEvent()} >Create</button>}
                        </section>
                        {
                            createNewEvent && 
                            <NewEventForm 
                                titleRef={titleRef}
                                infoRef={infoRef}
                                fromDateRef={fromDateRef}
                                toDateRef={toDateRef} 
                                onColorChange={handleColorChange}
                            />
                        }
                    </section>
                    <section className="calendar-menu-google">
                        <header className="calendar-menu-google-header">
                            <Icon src={"./google-calendar.svg"} alt={"Google Calendar Icon"} />
                            <h3>Google Calendar</h3>
                        </header>
                        <hr />
                        <p>Import your calendars and sync them to import their events.</p>
                        <button className="import-google-btn" onClick={getGoogleCalendars}>
                            <BsFillCalendarPlusFill />
                            Import calendars
                        </button>
                        <ul className="calendar-menu-google-calendars">
                        {
                            calendars.map((calendar, index) => {
                                return (
                                    <li key={index} className="calendar-item">
                                        <p style={{backgroundColor: calendar.color}}>{calendar.name}</p>
                                        <CalendarHandler calendarId={calendar.id} />
                                    </li>
                                )
                            })
                        }
                        </ul>
                    </section>
                </article>
            }
        </div>
    )
}

export interface CalendarProps {
    calendarId: string,
}
function CalendarHandler({calendarId}: CalendarProps) {

    const {events, setEvents} = useEventContext();

    /**
     * Fetch events of an specific calendar from Google Calendar. 
     * Makes a call to the API which responds with a list of events.
     */
    const getGoogleCalendarEvents = (calendarId:string) => {
        fetch('http://localhost:8080/google/events/get', {
            method: 'POST',
            credentials: 'include',
            headers: { 'Content-Type': 'application/json'},
            body: JSON.stringify({calendarId}),
        })
        .then(response => response.json())
        .then(data => {
            data.map((event:any) => {
                // the date coming from the backend is in format YYYY-MM-DDTHH:MM:SS.000Z => need to convert it to Fri Mar 29 ...
                // when doing just new Date(...) the timezone is applied, changing the datetime fetched from the backend.
                // The patch with the ISOString does not change datetimes.
                event.start = new Date(new Date(event.start).toISOString().slice(0, 16));
                event.end = new Date(new Date(event.end).toISOString().slice(0, 16));
            })
            setEvents((prev) => [...prev, ...data]);
        })
        .catch(error => console.error('Error:', error));
    }

    return (
        <button onClick={() => getGoogleCalendarEvents(calendarId)} className="calendar-item-sync-btn">
            <IoMdSync size={"1rem"} /> Synchronize
        </button>
    )
}