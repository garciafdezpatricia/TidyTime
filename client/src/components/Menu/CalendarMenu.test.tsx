import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import CalendarMenu from '@/src/components/Menu/CalendarMenu';

describe('DonutChart', () => {
    it('renders without crashing', () => {
      const { container } = render(<CalendarMenu />);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<CalendarMenu />);
        expect(asFragment()).toMatchSnapshot();
      });
  });