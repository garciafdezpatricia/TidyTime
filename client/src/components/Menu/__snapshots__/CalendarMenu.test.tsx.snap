// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChart matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      class="calendar-menu-icon"
      data-testid="calendar-menu-icon"
    >
      <svg
        fill="currentColor"
        height="1.4rem"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(0, 0, 0); background-color: transparent; border-radius: 0.4rem;"
        viewBox="0 0 512 512"
        width="1.4rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M88 152h336M88 256h336M88 360h336"
          fill="none"
          stroke-linecap="round"
          stroke-miterlimit="10"
          stroke-width="48"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
